src/wasm/wasm-s-parser.cpp

```diff
@@ -846,14 +853,22 @@ void SExpressionWasmBuilder::preParseHeapTypes(Element& module) {
 
   for (auto& pair : typeIndices) {
     auto name = pair.first;
-    auto type = types[pair.second];
+    auto index = pair.second;
+    auto type = types[index];
     // A type may appear in the type section more than once, but we canonicalize
     // types internally, so there will be a single name chosen for that type. Do
     // so determistically.
     if (wasm.typeNames.count(type) && wasm.typeNames[type].name.str < name) {
       continue;
     }
-    wasm.typeNames[type].name = name;
+    auto& currTypeNames = wasm.typeNames[type];
+    currTypeNames.name = name;
+    if (type.isStruct()) {
+      auto& currFieldNames = fieldNames[index];
+      for (auto& pair : currFieldNames) {
+        currTypeNames.fieldNames[pair.first] = pair.second;
+      }
```

Comment:
Can this just be something like `currTypeNames.fieldNames = fieldNames[index]`?

Reply:
Yes, I think you're right. We don't update on top of anything existing, and I doubt we'll need to, so we can just replace. Done.